toolkit/tools/imagegen/installutils/installutils.go

```diff
@@ -1452,6 +1458,102 @@ func updateUserPassword(installRoot, username, password string) (err error) {
 	return
 }
 
+// SELinuxConfigure pre-configures SELinux file labels and configuration files
+func SELinuxConfigure(systemConfig configuration.SystemConfig, installChroot *safechroot.Chroot, mountPointToFsTypeMap map[string]string) (err error) {
+	logger.Log.Infof("Preconfiguring SELinux policy in %s mode", systemConfig.KernelCommandLine.SELinux)
+
+	err = selinuxUpdateConfig(systemConfig, installChroot)
+	if err != nil {
+		logger.Log.Errorf("Failed to update SELinux config")
+		return
+	}
+	err = selinuxRelabelFiles(systemConfig, installChroot, mountPointToFsTypeMap)
+	if err != nil {
+		logger.Log.Errorf("Failed to label SELinux files")
+		return
+	}
+	return
+}
+
+func selinuxUpdateConfig(systemConfig configuration.SystemConfig, installChroot *safechroot.Chroot) (err error) {
+	const (
+		configFile     = "etc/selinux/config"
+		selinuxPattern = "^SELINUX=.*"
+	)
+	var mode string
+
+	switch systemConfig.KernelCommandLine.SELinux {
+	case configuration.SELinuxEnforcing, configuration.SELinuxForceEnforcing:
+		mode = configuration.SELinuxEnforcing.String()
+	case configuration.SELinuxPermissive:
+		mode = configuration.SELinuxPermissive.String()
+	}
+
+	selinuxConfigPath := filepath.Join(installChroot.RootDir(), configFile)
+	selinuxMode := fmt.Sprintf("SELINUX=%s", mode)
+	err = sed(selinuxPattern, selinuxMode, "`", selinuxConfigPath)
+	return
+}
+
+func selinuxRelabelFiles(systemConfig configuration.SystemConfig, installChroot *safechroot.Chroot, mountPointToFsTypeMap map[string]string) (err error) {
+	const (
+		squashErrors        = false
+		configFile          = "etc/selinux/config"
+		fileContextBasePath = "etc/selinux/%s/contexts/files/file_contexts"
+	)
+	var listOfMountsToLabel []string
+
+	// Search through all our mount points for supported filesystem types
+	// Note for the future: SELinux can support any of {btrfs, encfs, ext2-4, f2fs, jffs2, jfs, ubifs, xfs, zfs}, but the build system currently
+	//     only supports the below cases:
+	for mount, fsType := range mountPointToFsTypeMap {
+		switch fsType {
+		case "ext2", "ext3", "ext4":
+			listOfMountsToLabel = append(listOfMountsToLabel, mount)
+		case "fat32", "fat16", "vfat":
+			logger.Log.Debugf("SELinux will not label mount at (%s) of type (%s), skipping", mount, fsType)
+		default:
+			err = fmt.Errorf("Unknown fsType (%s) for mount (%s), cannot configure SELinux", fsType, mount)
+			return
+		}
+	}
+
+	// Find the type of policy we want to label with
+	selinuxConfigPath := filepath.Join(installChroot.RootDir(), configFile)
+	stdout, stderr, err := shell.Execute("sed", "-n", "s/^SELINUXTYPE=\\(.*\\)$/\\1/p", selinuxConfigPath)
+	if err != nil {
+		logger.Log.Errorf("Could not find an SELINUXTYPE in %s", selinuxConfigPath)
+		logger.Log.Error(stderr)
+		return
+	}
+	selinuxType := strings.TrimSpace(stdout)
```

Comment:
Is there a well-known list of supported SELINUXTYPE= options that we could check for?

Reply:
It corresponds to the `NAME` option in the selinux-policy compilation, which is an arbitrary string. There are common ones like `targeted`, `strict`, `mls`, `refpolicy`, and `minimum`, but it could be anything.

Reply:
Ah I see. Thanks for the explanation.